### PR TITLE
Add check point and log rotation metrics to Metrics plugin

### PR DESCRIPTION
**Pull request by davidegrohmann**

### CODE REVIEW

File: community/kernel/src/main/java/org/neo4j/kernel/impl/api/DefaultTransactionTracer.java [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+        @Override
+        public void setLogRotated( boolean logRotated )
+        {
+
+        }
+
+        @Override
+        public LogRotateEvent beginLogRotate()
+        {
+            startTimeNanos = clock.nanoTime();
```
Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/api/DefaultTransactionTracer.java:69]: imho it's overengineering...again, but feel free to keep it like it is

**Comment by benbc** [comment]
@davidegrohmann I don't see how these metrics are going to be usefully consumed. I'm going to carry on the discussion on the card rather than picking it up here.

**Comment by davidegrohmann** [comment]
@benbc PR updated by implementing the discussed solution

File: community/kernel/src/main/java/org/neo4j/kernel/impl/api/DefaultTransactionTracer.java [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+import org.neo4j.kernel.impl.transaction.tracing.TransactionEvent;
+import org.neo4j.kernel.impl.transaction.tracing.TransactionTracer;
+
+public class DefaultTransactionTracer implements TransactionTracer, LogRotationMonitor
+{
+    private final Clock clock;
+    private final AtomicLong counter = new AtomicLong();
+    private final AtomicLong accumulatedTotalTimeNanos = new AtomicLong();
+
+    private long startTimeNanos;
```
Comment (lutovich) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/api/DefaultTransactionTracer.java:42]: Field should probably be volatile

File: community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/log/checkpoint/DefaultCheckPointerTracer.java [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+import org.neo4j.kernel.impl.transaction.tracing.LogForceEvent;
+import org.neo4j.kernel.impl.transaction.tracing.LogForceWaitEvent;
+
+public class DefaultCheckPointerTracer implements CheckPointTracer, CheckPointerMonitor
+{
+    private final Clock clock;
+    private final AtomicLong counter = new AtomicLong();
+    private final AtomicLong accumulatedTotalTimeNanos = new AtomicLong();
+
+    private long startTimeNanos;
```
Comment (lutovich) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/log/checkpoint/DefaultCheckPointerTracer.java]: Checkpoint could be triggered from different threads, so I assume this field should be volatile

File: enterprise/metrics/src/main/java/org/neo4j/metrics/source/DBMetrics.java [hunk 4]
```diff
@@ -195,6 +210,50 @@ public Long getValue()
             } );
         }
 
+        // check pointing
+        if ( config.get( MetricsSettings.neoCheckPointingEnabled ) )
```
Comment (lutovich) [inline enterprise/metrics/src/main/java/org/neo4j/metrics/source/DBMetrics.java:214]: Method `start()` is quite big now. Could we please extract all `Gauge` setup into separate methods like `registerTxGauges`, `registerCheckPointGauges`, etc. ?

File: community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/log/checkpoint/DefaultCheckPointerTracer.java (hunk 3 above)
Comment (davidegrohmann) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/log/checkpoint/DefaultCheckPointerTracer.java]: Yes, but it is the same thread creating the `LogCheckPointEvent` and closing it. I can make that field volatile but maybe there is no need for it?

File: community/kernel/src/main/java/org/neo4j/kernel/impl/api/DefaultTransactionTracer.java (hunk 2 above)
Comment (davidegrohmann) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/api/DefaultTransactionTracer.java:42]: It is the same thread creating the `LogAppendEvent` and the `LogRotateEvent` and closing both. Is the `volatile` really needed?

File: enterprise/metrics/src/main/java/org/neo4j/metrics/source/DBMetrics.java (hunk 4 above)
Comment (davidegrohmann) [inline enterprise/metrics/src/main/java/org/neo4j/metrics/source/DBMetrics.java:214]: I have solved this problem in another PR: https://github.com/davidegrohmann/neo4j/commit/101b386926a419854808dd36538ac68777e8f3bd

**Pull request merged**